Process procs now properly utilize deltatime when implementing rates, timers and probabilities

Donkie opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Most of our codebase define rates and probabilities "per tick". Repaired health per tick, chance of breaking per tick, etc etc.

This is quite dumb since it means if we want to change the tickrate of things, we need to re-tune every single thing that uses that tickrate. We also have some code that is impacted by multiple tickrates which create even more complex dynamics. Changing the tickrate of systems should impact performance, stability and accuracy of the solution, not impact the state trajectory.

Another reason for this is that physical equations often require a time component. The volume pump has incorrectly been spitting 400L/s for years since the atmos tick is at 0.5s, yet the process proc for the pump makes no effort in treating this behaviour.

Some details about my implementation:
- My original plan was to send a "world.time - last_run" float as a parameter to the process proc, which would yield even more accurate physics simulations, and will also circumvent the issues that can arise when subsystems are ignored due to too much tick usage. However I decided to go with the #define macro route so that as many of these calculations will happen at compile time instead of runtime, to make sure that this PR doesn't impact performance.
- Most `prob` in process procs have been replaced with my homemade `DT_PROB` macro. It's explained in code, but it correctly calculates the adjusted probability based on different tickrates. The downside to this macro is that it requires a power operator so it can be expensive if done runtime, however most of the situations contain only hardcoded numbers and they are thus calculated at compiletime.
- Removed the `seconds` or `wait * 0.1` arguments that is sent in the process proc because literally nothing uses it.

If you think I should rename the constants to something else, or move their definitions somewhere else, please say so.

If this is appreciated, I will also deltatimefy the other subsystems like mobs.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Easier for people to tweak subsystem's tickrates in order to optimize server performance.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
code: Process procs now properly utilize deltatime when implementing rates, timers and probabilities, which paves way for easier subsystem tickrate adjustments.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Donkie committed:
Process procs now properly use deltatime when implementing rates, timers and probabilities

Qustinnus commented:
I've been waiting for something like this for quite some time. It'll probably go unnoticed but this'll make balancing way more easier if people can think in units per second instead of units per tick.

One big thing I noticed from looking at the code is that you didn't actually convert any of the numbers from tick to second, which means that everything that for example anything that runs in SSMOBS will now run twice as quick (As it ticks once every two seconds, but you now compensate for deltatime rather than just using tick). This'll need to be fixed by manually dividing all of the numbers by these static delta-times.

Qustinnus commented:
also tick defines should probably go in the _Defines/ file that has subsystem stuff

Donkie commented:
I actually have converted all numbers. Some DTs however are at 1 sec so they didn't need touching.

Alright I will move them

Qustinnus commented on code/modules/mob/living/silicon/pai/pai.dm:
```diff
@@ -314,7 +314,7 @@
 
 /mob/living/silicon/pai/process()
-	emitterhealth = clamp((emitterhealth + emitterregen), -50, emittermaxhealth)
+	emitterhealth = clamp((emitterhealth + emitterregen * SSMOBS_DT), -50, emittermaxhealth)
```
not this one

Qustinnus commented on code/modules/antagonists/changeling/powers/mutations.dm:
```diff
@@ -496,7 +496,7 @@
 	if(ishuman(loc))
 		var/mob/living/carbon/human/H = loc
-		H.reagents.add_reagent(/datum/reagent/medicine/salbutamol, REAGENTS_METABOLISM)
+		H.reagents.add_reagent(/datum/reagent/medicine/salbutamol, REAGENTS_METABOLISM * (SSOBJ_DT / SSMOBS_DT))
```
why would this one need 2 DTs? it affects mobs but it runs on the objects process() so you only need SSOBJ_DT

Donkie replied:
Since it uses reagents_metabolism constant I assume it's supposed to fill just enough for the mob to metabolize. Thus both DTs are needed.

Qustinnus replied:
right, makes sense.

Qustinnus commented:
my bad yeah I misread some code. this looks good then but i'm not sure about the need of using 2 deltatimes on one process(), can you explain why thats necesary?

Donkie commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -64,7 +65,7 @@
 
 /obj/effect/anomaly/proc/anomalyEffect()
-	if(prob(movechance))
+	if(DT_PROB(movechance, SSOBJ_DT))
```
Make macro const instead if this isn't changed anywhere

Donkie commented on code/game/objects/items/RSF.dm:
```diff
@@ -192,7 +192,7 @@ RSF
 /obj/item/rsf/cookiesynth/process()
-	matter = min(matter + 1, max_matter) //We add 1 up to a point
+	matter = min(matter += SSOBJ_DT, max_matter) //We add 1 up to a point
 	if(matter >= max_matter)
```
max_matter needs changing

Donkie commented on code/modules/clothing/spacesuits/hardsuit.dm:
```diff
@@ -77,14 +77,13 @@
 	grace_count = 0
-	rad_record -= rad_record/5
-	rad_record += rad_count/5
+	rad_record = (SSOBJ_DT / (8 + SSOBJ_DT)) * rad_count + (1 - SSOBJ_DT / (8 + SSOBJ_DT)) * rad_record // https://en.wikipedia.org/wiki/Low-pass_filter#Simple_infinite_impulse_response_filter
```
Make a macro for this instead and use that for the other Geiger counter as well

MrStonedOne replied:
the old code is basically just `MC_AVERAGE()`

The new code? fuck if i know.

Given the intention of this, I don't think you need to scale the math here, but if you did, the multiplication form would be more readable.

`rad_record = (rad_count * (0.20 * (SSOBJ_DT/2))) + (rad_record * (0.80 * (SSOBJ_DT/2)))`

Donkie replied:
It's a low pass filter, which is basically a moving average filter. I realise that this new form looks awful so I will make some macro that makes it more readable.

Donkie commented on code/modules/ninja/suit/n_suit_verbs/energy_net_nets.dm:
```diff
@@ -44,8 +44,8 @@ It is possible to destroy the net by the occupant or someone else.
-	if(check>0)
-		check--
+	if(check > 0)
+		check -= SSOBJ_DT // Not sure what is processing this, might be SSMACHINES_DT instead.
```
Confirm this bad boy

Donkie replied:
Mental note: /obj/structure/energy_net is not being processed at all

optimumtact commented:
@MrStonedOne

Qustinnus commented:
if this is merged I'll PR some guidelines for it in contributing.MD

optimumtact commented:
Most of our codebase define rates and probabilities "per tick". Repaired health per tick, chance of breaking per tick, etc etc.

This is quite dumb since it means if we want to change the tickrate of things, we need to re-tune every single thing that uses that tickrate.

Don't call it per tick, call it per cycle, that way people won't be confused between it and the world time ticking

MrStonedOne commented:
The entire system seems to be based on seconds, but all of byond and all our other systems are based on dediseconds. I foresee this causing issues and confusion, and I don't know that there is any benefit to doing it this way and creating two parallel units of times. 

@optimumtact what do you think about this? I feel like we should standardize on one or other and avoid systems that expect something different (outside of tick counts as a unit, for the rare thing that makes sense to scale based on game ticks). IIRC, we have a few cases of things that are per-second in code, and i remember many cases in old #coderbus of this confusing people.

The defines bit I'm also not sure about, for one, admins can change the fire rate of a system, I do often when debugging lag on the live servers. I get there is a cpu/speed component to the decision, but how much of a concern is that really? Such a system would work better and be easier for programmers to use if it just took a subsystem as an argument, and looked at that subsystem's actual wait value.   
(or even the delta between the last run and the current run (if you do this, you'll need to change master.dm:535 to use `queue_node.queued_time` instead of `world.time` for last_fire, and use `SS.queued_time-SS.last_fire` when calculating the delta, otherwise you'll compare the *start* of a subsystem's fire to when it last *ended*, skewing results on subsystems that take more then a game tick to run)).

> now properly utilize deltatime

This implies deltatime was a thing before, (and nothing was properly using it), but this is not true is it?

Also it seems like there should be some macros for math-operation-over-time. ie `SS_MATH(+, num, per-ds-change, SSblah)` (but less lame)

One last tip: byond will compile away any constant math operation inside of a (), and macro arguments that are constant count. it will try do this in other situations, but you can use `()` to be sure.

Qustinnus commented:
> This implies deltatime was a thing before, (and nothing was properly using it), but this is not true is it?


some process() procs had static deltatime as arg

optimumtact commented:
however, if we use deciseconds we align with world.time, and that might be less confusing

Donkie commented:
I had not considered the fact that you might want to tweak the subsystem wait on the fly. That could definitely be something that's worth embracing because it let's us test if things still work fine for longer tickrates. 

Deltatime is a concept that always should exist when dealing with game loops, because that is how you properly implement rate of change. And as @Qustinnus  said, someone made some feeble attempt to let everyone use Deltatime when calling `process()` but it seems like nobody took notice. 

Good point about the queued_time.

I'm not quite sure what you mean with the suggested SS_MATH macro
@MrStonedOne

If the defines were to use deciseconds it would just require me to shove ` SECONDS` after every case of using them, increasing bloat. Why not use seconds as a unit to begin with then?

Donkie committed:
Review fixes

Donkie committed:
Geiger counters cleanup

Made hardsuit geiger code more similar to geiger counter code
Geiger counters are more responsive now

Donkie committed:
Moved SS*_DT defines to subsystems.dm

Donkie commented:
Did some more changes. @MrStonedOne if you believe it's better to have a dynamic queued_time deltatime instead of this constant one system, just say and I'll go ahead and change it up. Most of the grunt work has been done already.

Donkie commented:
Sorry to bug you @MrStonedOne but what do you think?

optimumtact commented:
get this mergeable fyi, if he doesn't reply within a couple of days we'll just progress with this.

Donkie commented:
@optimumtact hmu when you want to merge this if it isn't merge-ready

MrStonedOne commented on code/__DEFINES/subsystems.dm:
```diff
@@ -237,3 +237,15 @@
 #define SSEXPLOSIONS_THROWS 3
 
+// Subsystem delta times or tickrates, in seconds. I.e, how many seconds in between each process() call for objects being processed by that subsystem.
+#define SSAIR_DT 0.5
+#define SSDCS_DT 1
+#define SSFASTPROCESS_DT 0.2
+#define SSMACHINES_DT 2
+#define SSMOBS_DT 2
+#define SSMOOD_DT 1
+#define SSNANITES_DT 1
+#define SSOBJ_DT 2
+#define SSPROCESSING_DT 1
+#define SSQUIRKS_DT 1
+#define SSRADIATION_DT 1
```
change all of this to use the wait. 

```dm
#define SSAIR_DT (SSair.wait/10)
...
```

I don't like duplicating info in a way that can create a desync, or making it so nobody can change a wait without changing the same value in multiple places, and we do still need to support var edited wait times.

I might also suggest adding SECONDS to the define, to avoid confusion (such as `SSAIR_DT_SECONDS`), but don't feel like blocking on that.

Donkie replied:
The waits have been updated to use these defines, so it's only defined in one place.

MrStonedOne replied:
Thats not what i asked.

The waits should remain dediseconds, both in given unit and stored unit.

Changing them to use these defines doesn't help with varedited waits. 

Change these defines to refer to the subsystem wait, keep the subsystem wait defined in the subsystem's file.

Donkie replied:
Oh I see what you mean, then we will lose the benefit of having this available at compile time. If we're going that route I'd prefer if it was actually dynamic like we discussed with queued_time etc.

MrStonedOne replied:
it won't quite work as i expected, some procs and sleep and resume after fire() has ended, the calculation for those cases will be wildly off. my plan is to go through and use those defines to quickly make everything use it.

and for the reference, subsystems will exist before any object code runs. (outside of stat()), so i don't forsee an issue to just making those defines refer to the wait by global var reference.

Donkie replied:
Alright, I'll convert the defines then

Donkie committed:
Rebase fix

Donkie committed:
Redefined the SS*_DT defines to use the subsystem wait vars

Donkie commented:
@MrStonedOne implemented your suggestions

AnturK commented on code/game/objects/effects/alien_acid.dm:
```diff
@@ -73,9 +73,9 @@
+		if(DT_PROB(25, SSOBJ_DT))
 			playsound(loc, 'sound/items/welder.ogg', 100, TRUE)
-		target_strength--
+		target_strength -= SSOBJ_DT * 0.5
```
Probably round given the switch below. (or changing that)

Donkie replied:
Done

AnturK commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -363,7 +365,7 @@
 		else
 			cyborg.cell.use(5)
-		repair_tick = 0
+		next_repair = world.time + repair_cooldown SECONDS
```
Using these defines look pretty bad outside of simple defines to me. Not very helpful with understanding the meaning either since the result is not in seconds.

Donkie replied:
Alright, fair enough

AnturK commented on code/game/turfs/open/lava.dm:
```diff
@@ -137,9 +137,9 @@
-				O.armor = O.armor.setRating(fire = 50)
-			O.fire_act(10000, 1000)
+			if(O.armor.fire > 50 * delta_time) //obj with 100% fire armor still get slowly burned away.
+				O.armor = O.armor.setRating(fire = 50 * delta_time)
```
That doesn't make much sense, whatever the armor was it was cut down to 50 before ?

Donkie replied:
You're right, that's pretty weird.

AnturK reviewed (commented):
Good PR, you should add a readme string to base process explaining the use of these defines.

Donkie committed:
Implemented suggested changes by @AnturK

Donkie commented on code/game/machinery/slotmachine.dm:
```diff
@@ -64,7 +64,7 @@
 		return .
 
-	money++ //SPESSH MAJICKS
+	money += SSMACHINES_DT / 2 //SPESSH MAJICKS
```
Done

Donkie committed:
Commented /datum/proc/process about the deltatime stuff

Donkie committed:
Send delta_time as a process parameter instead of the defines

Also DTfied acid_processing

Donkie committed:
Dtfied new acid component

Donkie commented:
@optimumtact merge pls

AnturK commented:
Going to testmerge for a moment and if nothing breaks obviously i'll merge.

AnturK commented:
Seems there's no major issues so good to go when conflicts are fixed.

Donkie commented:
@AnturK Fixed conflicts
